feat: add db service

Extheoisah opened the pull request.

Extheoisah committed:
feat: add diesel config and user schema

Extheoisah committed:
fix: nix flakes postgres deps issue when destructuring text

Extheoisah committed:
docs: add contributing guide

Extheoisah committed:
feat: add respository for creating and getting users

Extheoisah committed:
feat: add challenge repository

Extheoisah committed:
feat(refactor): refactor user create model and function

Extheoisah committed:
feat: add exercise repository

Extheoisah committed:
feat(refactor): move model impl to repository

Extheoisah committed:
feat: add progress repository

Extheoisah committed:
feat(refactor): add status to exercise schema and update status type in primitives & repositories

Extheoisah committed:
feat: add repository and submissions to db schema

Extheoisah committed:
feat: add session repository

Extheoisah committed:
feat: add leaderboard, repo and bages repository

daviroo commented on migrations/2024-08-28-124527_add-repo-submissions-table/down.sql:
```diff
@@ -0,0 +1,4 @@
+-- This file should undo anything in `up.sql`
+DROP TABLE submissions;
+DROP TABLE repositories;
```
```suggestion
DROP TABLE repositories;
DROP TABLE submissions;
```

daviroo replied:
The order should be switched since `submissions` has a dependency on `repositories`.

Extheoisah replied:
Good catch!

Extheoisah replied:
addressed here -> 1deec9aca94e30d3d33a26f8cbe5207e3dd99d8d

daviroo commented on migrations/2024-08-28-124527_add-repo-submissions-table/up.sql:
```diff
@@ -0,0 +1,25 @@
+    status VARCHAR(255) NOT NULL CHECK (status IN ('pending', 'passed', 'failed')),
+    feedback TEXT,
+    submitted_at TIMESTAMP NOT NULL DEFAULT now(),
+    updated_at TIMESTAMP NOT NULL DEFAULT now()
```
This will be set to datetime at time of row creation and will need to be updated manually on every update after that (unless we add a database trigger).  Just something to note.

Extheoisah replied:
Diesel automatically added a trigger SQL script that automatically updates the `updated_at` field whenever there's an update to the record which was why I did add the trigger in the migration scripts. see here -> https://github.com/Extheoisah/hxckr-core/pull/10/files#diff-79ed83666bddb98b5ca0f9f9f009f0208d86e218a232fea1f4f6dc503b2b9b5bR19

Extheoisah replied:
I tested that this trigger works. See timestamp in attached screenshot
<img width="215" alt="Screenshot 2024-08-30 at 14 35 56" src="https://github.com/user-attachments/assets/276356dd-d510-4deb-8769-eb555e6feae2">

daviroo replied:
I think you need to use the trigger function and reference the table name for it to work?  I could be wrong.  There's an example in the comments above that function you linked to.

Extheoisah replied:
Actually, during my testing, I didn't include it anywhere. I just updated the leaderboard record and the updated at field was automatically updated. I see from my db inspector that the trigger function is already added automatically by diesel
<img width="1005" alt="Screenshot 2024-08-30 at 14 39 35" src="https://github.com/user-attachments/assets/60f2fd14-df28-4292-9c6f-017f98c8d16f">

Extheoisah replied:
So I cross-checked again to be sure that I was correct, but I was wrong, and you're right. The function wasn't automatically applied, and the reason I saw that the leaderboard record was updated was that I was manually updating the updated_at field in the update function.
I have added a migration script to reference the diesel_manage_updated_at trigger function for all tables except for the session table. 
Sorry for the mix-up. See changes here -> https://github.com/Extheoisah/hxckr-core/pull/10/commits/c2f5744b467abe37a75c031bedbf54a2939c46ca

daviroo commented on migrations/2024-08-27-153454_create-challenges-table/up.sql:
```diff
@@ -0,0 +1,11 @@
+    difficulty VARCHAR(255) CHECK (difficulty IN ('easy', 'medium', 'hard')) NOT NULL,
+    mode VARCHAR(255) CHECK (mode IN ('project', 'functional_test')) NOT NULL,
+    created_at TIMESTAMP NOT NULL DEFAULT now(),
+    updated_at TIMESTAMP NOT NULL DEFAULT now()
```
This will be set to datetime at time of row creation and will need to be updated manually on every update after that (unless we add a database trigger).  Just something to note.

daviroo commented on migrations/2024-08-27-184211_create-exercises-table/up.sql:
```diff
@@ -0,0 +1,13 @@
+    test_runner VARCHAR(255) NOT NULL,
+    challenge_id UUID NOT NULL REFERENCES challenges(id),
+    created_at TIMESTAMP NOT NULL DEFAULT now(),
+    updated_at TIMESTAMP NOT NULL DEFAULT now()
```
This will be set to datetime at time of row creation and will need to be updated manually on every update after that (unless we add a database trigger).  Just something to note.

daviroo commented on migrations/2024-08-28-101632_create-progress-table/up.sql:
```diff
@@ -0,0 +1,13 @@
+    status VARCHAR NOT NULL CHECK (status IN ('completed', 'in_progress', 'not_started')) DEFAULT 'not_started',
+    progress_details JSONB,
+    created_at TIMESTAMP NOT NULL DEFAULT now(),
+    updated_at TIMESTAMP NOT NULL DEFAULT now()
```
This will be set to datetime at time of row creation and will need to be updated manually on every update after that (unless we add a database trigger).  Just something to note.

daviroo commented on migrations/2024-08-29-132507_add-leaderboards-table/up.sql:
```diff
@@ -0,0 +1,29 @@
+    achievements JSONB,
+    score INT NOT NULL DEFAULT 0,
+    created_at TIMESTAMP NOT NULL DEFAULT now(),
+    updated_at TIMESTAMP NOT NULL DEFAULT now()
```
This will be set to datetime at time of row creation and will need to be updated manually on every update after that (unless we add a database trigger).  Just something to note.

daviroo commented on src/service/repository/leaderboard.rs:
```diff
@@ -0,0 +1,110 @@
+            diesel::update(leaderboard_table.filter(user_id_col.eq(user_id_uuid)))
+                .set((
+                    score_col.eq(new_score),
+                    achievements_col.eq(serde_json::to_value(new_achievements).unwrap()),
```
I think if `new_achievements` is of value `None` here then the `achievements` column will be overwritten with `null`.

Extheoisah replied:
Hmm... you have a point here. I'll check and revert.

Extheoisah replied:
resolved here https://github.com/Extheoisah/hxckr-core/pull/10/commits/f535313fe5c2040f9f75eea85a3daeb4e8149e27

daviroo reviewed (changes requested):
Mostly small stuff but one potential bug in the `leaderboard` repository.

daviroo commented on migrations/00000000000000_diesel_initial_setup/up.sql:
````diff
@@ -0,0 +1,36 @@
+-- ```sql
+-- CREATE TABLE users (id SERIAL PRIMARY KEY, updated_at TIMESTAMP NOT NULL DEFAULT NOW());
+--
+-- SELECT diesel_manage_updated_at('users');
````
I think this might be an example of how to use the provided trigger to automatically populated the `updated_at` columns without us having to do it manually.

Extheoisah committed:
feat: add badge and user badge repository

Extheoisah committed:
fix: switch order of repo-submission drop migration script

addresses review comment https://github.com/Extheoisah/hxckr-core/pull/10#discussion_r1738650374

Extheoisah committed:
fix: only update specified leaderboard fields

Extheoisah committed:
feat: add updated_at trigger function to migration script
